client/starwhale/api/_impl/dataset.py

```diff
@@ -24,6 +36,132 @@
+        self.object_store_type = object_store_type
+        self.label = label.encode() if isinstance(label, str) else label
+
+        # TODO: add non-starwhale object store related fields, such as address, authority
+        # TODO: add data uri crc for versioning
+        # TODO: support user custom annotations
+
+    def _do_validate(self) -> None:
+        if self.id < 0:
+            raise FieldTypeOrValueError(
+                f"id need to be greater than or equal to zero, but current id is {self.id}"
+            )
+
+        if not self.data_uri:
+            raise FieldTypeOrValueError("no raw_data_uri field")
+
+        if self.data_format not in [RawDataFormatType.SWDS_BIN, RawDataFormatType.USER]:
+            raise NoSupportError(f"data format: {self.data_format}")
+
+        if self.data_origin not in [DataOriginType.NEW, DataOriginType.INHERIT]:
+            raise NoSupportError(f"data origin: {self.data_origin}")
+
+        # TODO: support non-starwhale remote object store, for index-only feature
+        if self.object_store_type != ObjectStoreType.LOCAL:
+            raise NoSupportError(f"object store {self.object_store_type}")
+
+    def __str__(self) -> str:
+        return f"row-{self.id}, data-{self.data_uri}, origin-[{self.data_origin}]"
+
+    def __repr__(self) -> str:
+        return f"row-{self.id}, data-{self.data_uri}(offset:{self.data_offset}, size:{self.data_size}, format:{self.data_format}), origin-[{self.data_origin}], object store-{self.object_store_type}"
+
+    def asdict(self) -> t.Dict[str, t.Union[str, bytes, int]]:
+        return deepcopy(self.__dict__)
+
+
+class TabularDataset:
+    def __init__(self, name: str, version: str, project: str) -> None:
+        self.name = name
+        self.version = version
+        self.table_name = f"{name}/{version[:VERSION_PREFIX_CNT]}/{version}"
+        self._ds_wrapper = DatastoreWrapperDataset(self.table_name, project)
+
+        self.start = -1
+        self.end = sys.maxsize
+
+        self._do_validate()
+
+    def _do_validate(self) -> None:
+        _ok, _reason = validate_obj_name(self.name)
+        if not _ok:
+            raise InvalidObjectName(f"{self.name}: {_reason}")
+
+        if not self.version:
+            raise FieldTypeOrValueError("no version field")
+
+    def __str__(self) -> str:
+        return f"Dataset Table: {self._ds_wrapper._meta_table_name}"
+
+    __repr__ = __str__
+
+    def set_range(self, start: int, end: int) -> None:
+        self.start = max(start, self.start)
+        self.end = min(end, self.end)
+
+    def put(self, row: TabularDatasetRow) -> None:
+        self._ds_wrapper.put(row.id, **row.asdict())
+
+    def scan_all(self) -> t.Generator[TabularDatasetRow, None, None]:
+        return self.scan(self.start, self.end)
+
+    def scan(self, start: int, end: int) -> t.Generator[TabularDatasetRow, None, None]:
+        for _d in self._ds_wrapper.scan(start, end):
+            yield TabularDatasetRow(**_d)
```

Comment:
When a function accepts a  `TabularDataset` with a predefined range (e.g. called `set_range(2, 4)` outside).
Calling dataset.scan_all() will get [2, 3], and dataset.scan(0, 2) will get [0, 1].
It doesn't feel as expected.

Reply:
fixed